pkg/clustermesh/common/namespacewatcher/watcher.go

```diff
@@ -0,0 +1,480 @@
+	// GetAllNamespaces returns all namespaces that have resources in this processor's resource index
+	GetAllNamespaces() []string
+}
+
+const (
+	// OptClusterMeshDefaultGlobalNamespace is the name of the clustermesh-default-global-namespace option
+	OptClusterMeshDefaultGlobalNamespace = "clustermesh-default-global-namespace"
+)
+
+// Config configures the namespace watcher behavior.
+type Config struct {
+	// DefaultGlobalNamespace determines the default behavior for namespaces when
+	// namespace-based filtering is active (i.e., when at least one namespace is annotated).
+	// When true, namespaces are global by default unless annotated with clustermesh.cilium.io/global=false.
+	// When false, namespaces are local by default unless annotated with clustermesh.cilium.io/global=true.
+	DefaultGlobalNamespace bool `mapstructure:"clustermesh-default-global-namespace"`
+}
+
+// Flags implements cell.Flagger to register the clustermesh-default-global-namespace flag.
+func (cfg Config) Flags(flags *pflag.FlagSet) {
+	flags.Bool(OptClusterMeshDefaultGlobalNamespace, cfg.DefaultGlobalNamespace,
+		"Default behavior for namespaces when namespace-based filtering is active. "+
+			"When true, namespaces are global by default unless annotated with 'clustermesh.cilium.io/global=false'. "+
+			"When false, namespaces are local by default unless annotated with 'clustermesh.cilium.io/global=true'.")
+}
+
+type namespaceWatcher struct {
+	logger            *slog.Logger
+	config            Config
+	mu                lock.RWMutex
+	namespaceResource resource.Resource[*slim_corev1.Namespace]
+
+	// Processors for handling namespace changes
+	processors []NamespaceProcessor
+}
+
+// NewNamespaceWatcher creates a new namespace watcher that tracks global namespaces.
+func NewNamespaceWatcher(logger *slog.Logger, config Config, namespaceResource resource.Resource[*slim_corev1.Namespace]) *namespaceWatcher {
+
+	return &namespaceWatcher{
+		logger:            logger,
+		config:            config,
+		namespaceResource: namespaceResource,
+	}
+}
+
+func (nw *namespaceWatcher) RegisterProcessor(processor NamespaceProcessor) {
+	nw.mu.Lock()
+	defer nw.mu.Unlock()
+	nw.processors = append(nw.processors, processor)
+}
+
+// getAllNamespacesFromProcessors gets all namespaces from all registered processors
+func (nw *namespaceWatcher) getAllNamespacesFromProcessors() []string {
+	allNamespaces := sets.New[string]()
+
+	// Collect namespaces from all processors
+	for _, processor := range nw.processors {
+		namespaces := processor.GetAllNamespaces()
+		for _, ns := range namespaces {
+			allNamespaces.Insert(ns)
+		}
+	}
+
+	return allNamespaces.UnsortedList()
+}
+
+// isNamespaceAnnotated checks if a namespace has the global annotation by querying the resource store
+func (nw *namespaceWatcher) isNamespaceAnnotated(namespace string) bool {
+	ctx := context.Background()
+	store, err := nw.namespaceResource.Store(ctx)
+	if err != nil {
+		return false
+	}
```

Comment:
Having this store retrieval step in all getters looks definitely a code smell to me, especially because then given a background context that could lead to methods being unexpectedly blocked forever. The store should be retrieved once (e.g., in the `namespace-watcher` job), and then simply accessed.

Reply:
You are right. Fixed.